Hide 'center' column when it's a result of an AOI analysis

rjimenezda opened the pull request:
Fix https://github.com/CartoDB/support/issues/1404
Fix https://github.com/CartoDB/support/issues/1149

AOI is adding an extra column `center` which is confusing. Since we're already hiding `the_geom_webmercator` we could just hide all geometry columns that are not `the_geom`

If we want to hide geometry columns called `center` but only when there's this type of analysis added, that's a whole different can of worms :melon:

andy-esch commented:
Sorry to poke my nose into something I haven't worked on, but I think hiding geometry columns could be pretty confusing if a user specifically creates a geometry column for use in their application. I agree with hiding `the_geom_webmercator` since it's a utility column, but not all geom columns other than `the_geom` are for the utility of the platform like that.

andy-esch commented:
Could there be a middle-ground solution of having a button that is 'expand to show hidden columns' instead of outright hiding them with no non-sql way of getting them to display in the table view?

ramiroaznar commented:
@andy-esch @zingbot Thanks for the feedback. I think we are covering your use cases, we are not going to hide any geometry column if there is a custom query applied. Are we missing something? If so, I recommend to try the fix in the following staging environment and give us the acceptance.

rjimenezda commented:
The only case we're not covering is if there's another geometry column called **center** (that's the only one we're hiding for now) from the start, without making a custom query.

This is entirely possible for people using the SQL API extensively, since you can add custom geometry columns using postgis.

So yeah, this is quite a pickle. These edge cases makes me wonder if we should alter this at all. The safest bet would be to hide stuff only if there's a time buffer analysis added (the distance buffer does not create the new column), but I'm not a big fan of that.

Andy's proposal is great, but it's not straightforward. We could also style the columns that are added by analyses differently (icon, some tint), to mitigate confusion.

We could also show `cartodb_id` & `the_geom` always before than the rest of the columns. I think most of the confusion of this issue is that `center` shows up before `the_geom` on the table view :man_shrugging:

ramiroaznar commented:
> So yeah, this is quite a pickle. These edge cases makes me wonder if we should alter this at all. The safest bet would be to hide stuff only if there's a time buffer analysis added (the distance buffer does not create the new column), but I'm not a big fan of that.

Let's do this. Let's just not show `center` in the layer view and also remove it from the export.

> Andy's proposal is great, but it's not straightforward. We could also style the columns that are added by analyses differently (icon, some tint), to mitigate confusion.

It is great. But this is the RT, we can add some enhancements but this is something out of the scope of this issue.

rjimenezda commented:
Staging username: hide-center-column

Now we're checking that there's a buffer time analysis on the layer and that there's **not** a SQL Analysis, since that one could add arbitrary new columns.

This should cover all cases for the `center` geometry column, which is the only one we're hiding (I'm going to rename the PR to avoid confusion).

If the user has a dataset that originally contains a `center` column (of any type), and they apply a time buffer analysis, this column **will be overwritten** with the result of the analysis.

I also thought that a JOIN analysis might cause problems, but they append right_ or left_ to the column name, so not an issue.

If there's any more cases, now'd be a great time to know :grimacing:

rjimenezda commented:
Redeploying - it was missing some master stuff and making it hard to check the data panel. Should be ready in ~30 mins from this message

rjimenezda commented:
We're only hiding geometry columns named `center` if there's an AOI analysis present and no SQL analysis present.

rjimenezda commented:
@VictorVelarde I've added some specs as well, so all should be good in da hood to merge on monday morning while I'm far, far away from the computer :grimacing: 

I haven't added any specs to the analysis utils that I've added, but they are quite straightforward, and there are specs testing their 'integration' within other places.

It would be nice if someone would take a look for CR, just in case I did something barbaric.

I'm going to rebase to cleanup all the 'bump version' commits, so don't get this branch yet :)

VictorVelarde commented:
retest this please

VictorVelarde commented:
After talking with #backend it seems we can ignore temporary Backend failing tests at Jenkins

VictorVelarde commented:
This has been redeployed to staging (user: hide-center-column). Could you please have a look at it @ramiroaznar ?

jsanz commented:
@oriolbx I don't have vpn credentials (I asked for them already), can you give it a try when you have some time please? :pray:

jsanz commented:
I'm having a hard time to test this, I've uploaded a minimal dataset `test_1` but the dataset view and the dashboard are firing this error

![2018-07-25_18 00 18-selection](https://user-images.githubusercontent.com/188264/43212680-f294164a-9034-11e8-9246-87a3c303ac5e.png)

I managed to try it once with the `populated_places` dataset and yea, the center column was not there but then I wanted to try with this simpler dataset and I'm not able to even load the map, I think is something related with the deployment because this url is not working

`{STAGING_URL}/api/v2/sql?q=SELECT+USER`

VictorVelarde commented:
It seems that staging env was somehow corrupted. 
I've redeployed to a new one: https://extra-column-aoi.carto-staging.com/

Could you please @jsanz ?

jsanz commented:
all good, no `center` column on the analysis output

![2018-07-26_11 43 30-selection](https://user-images.githubusercontent.com/188264/43255091-2ce945e2-90c9-11e8-952d-32c9191a049f.png)

:+1:

rjimenezda commented on lib/assets/test/spec/builder/components/modals/export-data/modal-export-data-view.spec.js:
```diff
@@ -150,7 +150,7 @@ describe('components/modals/export-data/modal-export-data-view', function () {
       describe('if the view can hide columns', function () {
         beforeEach(function () {
-          this.querySchemaModel.getColumnNames = () => ['center'];
+          this.querySchemaModel.getColumnNames = function () { return ['center']; };
```
Hey, so this was working on CI, so I'm wondering why it didn't work locally?

VictorVelarde replied:
That's true, it was, but double-checked locally and it failed.

rjimenezda replied:
So weird, I'll take a look eventually™️
